### ISSUE TITLE
feat: Custom threshold in SmartCorrelatedFeatures

### ISSUE THREAD

**Issue by TremaMiguel** [solegalli/feature_engine#287]
**Is your feature request related to a problem? Please describe.**

Currently `SmartCorrelatedFeatures` takes correlation measures that have a similar range (between -1 and +1) and select features by a fixed threshold value (defaults to 0.8). 

**Describe the solution you'd like**

Extend the `threshold` selection to a custom function in order to be able to apply other correlation measures that have different range as mutual information.  Other measures to consider could be find [here](https://m-clark.github.io/docs/CorrelationComparison.pdf).

**Describe alternatives you've considered**
An alternative is to create a separate class.

**Comment by solegalli** [comment]
I had a quick look at the link, and all metrics seem to return values somewhere between -1 and 1. 

Why do you think we need to expand the threshold allowance in the transformer?

Also, there is one metric, the Hoeffner D or similar, that varies between -0.5 and 1, and not sure how this metric works, but the transformer is taking the absolute value, so not sure how to make this one fit in, or even if it is a good idea to make it fit in.

**Comment by TremaMiguel** [comment]
For the MI calculation, I'm basing on [sklearn.feature_selection.mutual_info_regression](https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.mutual_info_regression.html). Basically uses [scipy.special.digamma](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.digamma.html) which is in log scale. In theory, there would be no upper bound for this measure. For two continuous variables, [here](https://github.com/scikit-learn/scikit-learn/blob/2beed55847ee70d363bdbfe14ee4401438fba057/sklearn/feature_selection/_mutual_info.py#L71) is how it is calculated.

The idea to modify this class is to return to the user the variables that are informative with MI. I think this could happen by passing a custom function to `threshold` that let user define how to select the features after `pandas.corr()` is calculated. Maybe passing custom method [here](https://github.com/solegalli/feature_engine/blob/master/feature_engine/selection/smart_correlation_selection.py#L255), not only removing features with high correlation.

**Comment by solegalli** [comment]
For the standard correlation functions (spearman, pearson, kendall) the value of correlation is quite intuitive, if it is >0.7 or 0.8, then the variables are probably very related.

What sort of value indicates correlation for this mutual_info_regression? Is this an accepted metric for correlation? where has it been used? (other than in sklearn, which I understand what a workaround to extend the MI to continuous targets)

**Comment by TremaMiguel** [comment]
The MI can detect non-linear, sinusoidal and linear relationships as well. [Here](https://www.kaggle.com/ryanholbrook/mutual-information) is a better explanation from a Kaggle Tutorial.

**Comment by solegalli** [comment]
In the kaggle kernel you mention it is used exactly as per sklearn docs. To determine relationship with the target and not among variables. It also does not make reference to which would be an accepted threshold to indicate that above the threshold the variables are "truly" related.

We need more info on the threshold front, and on how widely this metric is used to determine relationship among variables, instead of variables and target.

**Comment by TremaMiguel** [comment]
Yes, that's a good point, from what I know. Relations above 1 are periodic. 

```python
import numpy as np
import matplotlib.pyplot as plt

from sklearn.feature_selection import mutual_info_regression

n = 100
x = np.linspace(-4, 4, num=n)
y = np.sin(x) 

plt.plot(x, y)
mutual_info_regression(x.reshape(-1, 1), y.reshape(-1, 1))
>>> 1.32
````
![Captura de pantalla (111)](https://user-images.githubusercontent.com/49818188/126075363-c160901f-b0ef-4509-be27-a875483c9a9d.png)

Convex or non linear relations are close to two 

```python
x = np.linspace(-1, 4, num=n)
y = np.sin(x) 
mutual_info_regression(x.reshape(-1, 1), y.reshape(-1, 1))
>>> 1.8
````
![Captura de pantalla (112)](https://user-images.githubusercontent.com/49818188/126075438-01b4b702-28f1-4653-ba89-504e8ae3f899.png)

Monotonic relations above two 

```python
x = np.linspace(-4, 1, num=n)
y = np.exp(x) 

plt.plot(x, y)
mutual_info_regression(x.reshape(-1, 1), y.reshape(-1, 1))
>>> 2.05
```

![Captura de pantalla (113)](https://user-images.githubusercontent.com/49818188/126075499-bb81cecc-6624-4db1-abeb-ae1be5ab16b8.png)


and linear relations closer to 3

```python
import seaborn as sns
x = np.array([np.random.rand() * 100 for _ in range(100)])

y = x * 5 + np.random.randn(n) * 2

sns.scatterplot(x, y)
mutual_info_regression(x.reshape(-1, 1), y.reshape(-1, 1))
>>> 3.12
```

![Captura de pantalla (114)](https://user-images.githubusercontent.com/49818188/126075526-048e91b9-1e29-4fc7-b673-2b513461da84.png)


I know it is not yet widely adopted by the community, but I think it is a great addition to your toolkit to quickly understand the general relationship between the target and the covariates. 

There would be no difference to apply it to determine the relationship between variables, as the same function is always applied,  [here](https://github.com/scikit-learn/scikit-learn/blob/2beed55847ee70d363bdbfe14ee4401438fba057/sklearn/feature_selection/_mutual_info.py#L147).

In general, there are no many examples of its use, I think a general acceptance is values above zero, but not sure what relationship is detected for value between zero and one.

> It is equal to zero if and only if two random variables are independent, and higher values mean higher dependency.

An [example](https://scikit-learn.org/stable/auto_examples/feature_selection/plot_f_test_vs_mi.html#sphx-glr-auto-examples-feature-selection-plot-f-test-vs-mi-py) from sklearn.

**Comment by solegalli** [comment]
I hear the pros:
a tiny code modification,  that is, removing the check for the threshold value, would allow the use of this technique, and virtually any other.

On the downside, not having the threshold check, may encourage users to use any value, so for users less familiar with correlation, we would not be guiding them, when setting up the transformer, to choose a value that is suitable, for the more used correlation metrics.

If this technique was widely used, I would say, yes, let's remove the threshold check. But since we are not aware that this is the case, I am disinclined to proceed.

Has the mutual information been used at all to select features? maybe we need to explore more this, and if yes, we create a dedicated trasformer, to keep things simple and clear. The smartcorrelation selects based on correlation, and then, if suitable, we would have a selector based on mutual info. But I have not heard of this so far

**Comment by TremaMiguel** [comment]
Regarding 

> On the downside, not having the threshold check, may encourage users to use any value, so for users less familiar with correlation, we would not be guiding them, when setting up the transformer, to choose a value that is suitable, for the more used correlation metrics.

We notice that method `SmartCorrelatedSelection` is highly dependent on three correlation measures (spearman, kendall and pearson) this turns out the method inflexible to consider other correlation measures.

So as you say, we can tackle this with two approaches:

1. Modify `SmartCorrelatedSelection`  to remove that high dependency and make it flexible for other methods. (It already accepts any kind of function, we just need to extend the selection or threshold criteria). I think threshold can be a method appropiate to the range of the correlation measure. 

* check_zero_one: For correlation measures range ranging from zero to one.
* check_zero_inf: For correlation measures range ranging from zero to inf (mutual information).

any other ideas ? 

2. Separate transformer with Mutual Information method. This can be an experiment feature that could be further integrate into `SmartCorrelatedSelection`, it will basically adapt the behavior of `SmartCorrelatedSelection` to Mutual Information.

and regarding 

> Has the mutual information been used at all to select features? maybe we need to explore more this, and if yes, we create a dedicated trasformer, to keep things simple and clear. The smartcorrelation selects based on correlation, and then, if suitable, we would have a selector based on mutual info. But I have not heard of this so far

From my own experience using this method, I see this as an easy and quick way to detect relationships, If we offer such a method with a quick and easy way to calculate relations, this could have an impact on the visibility of feature-engine, also see it the only library currently offering this method.

**Comment by TremaMiguel** [comment]
@solegalli , can we decide on which of the issues #287, #290, #291, #292 and #293 to develop?

**Comment by solegalli** [comment]
Hi @TremaMiguel

Would you mind please removing the last comment? I don't think it is relevant to this issue. It is linking other issues that are not related.

Thank you!
